File: src/main/scala/com/michalplachta/galactic/service/TweetsService.scala
package com.michalplachta.galactic.service

import com.michalplachta.galactic.db.DbClient
import com.michalplachta.galactic.logic.TweetCensorship.censorTweetsUsingFilters
import com.michalplachta.galactic.values.RemoteData.{ Failed, Fetched, Loading, NotRequestedYet }
import com.michalplachta.galactic.values.{ RemoteData, Tweet }

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future
import scala.util.{ Failure, Success }

object TweetsService {
  private var cachedTweets: Map[String, RemoteData[List[Tweet]]] = Map.empty

  def getTweetsFor(citizenName: String): RemoteData[List[Tweet]] = {
    if (cachedTweets.get(citizenName).isEmpty) cachedTweets += (citizenName → Loading())
    getTweetsAsync(citizenName)
      .map(censorTweetsUsingFilters)
      .onComplete { triedTweets ⇒
        val remoteTweets: RemoteData[List[Tweet]] =
          triedTweets match {
            case Success(newTweets) ⇒ Fetched(newTweets)
            case Failure(t)         ⇒ Failed(t.toString)
          }
        cachedTweets += (citizenName → remoteTweets)
      }

    cachedTweets.getOrElse(citizenName, NotRequestedYet())
  }

  private def getTweetsAsync(citizenName: String): Future[List[Tweet]] = {
    for {
      citizen ← DbClient.findCitizenByName(citizenName)
      tweets ← DbClient.getTweetsFor(citizen)
    } yield tweets
  }
}

File: src/main/java/com/michalplachta/galactic/java/service/TweetsService.java
package com.michalplachta.galactic.java.service;

import com.michalplachta.galactic.java.db.DbClient;
import com.michalplachta.galactic.java.logic.TweetCensorship;
import com.michalplachta.galactic.java.values.remotedata.*;
import com.michalplachta.galactic.java.values.Citizen;
import com.michalplachta.galactic.java.values.Tweet;
import io.vavr.collection.HashMap;
import io.vavr.collection.List;
import io.vavr.collection.Map;
import io.vavr.concurrent.Future;

import static io.vavr.API.*;
import static io.vavr.Patterns.$Failure;
import static io.vavr.Patterns.$Success;

public class TweetsService {
    private static Map<String, RemoteData<List<Tweet>>> cachedTweets = HashMap.empty();

    public static RemoteData<List<Tweet>> getTweetsFor(String citizenName) {
        if (cachedTweets.get(citizenName).isEmpty()) cachedTweets = cachedTweets.put(citizenName, new Loading<>());
        getTweetsAsync(citizenName)
                .map(TweetCensorship::censorTweetsUsingFilters)
                .onComplete(triedTweets -> {
                    RemoteData<List<Tweet>> remoteTweets = Match(triedTweets).of(
                            Case($Success($()), Fetched::new),
                            Case($Failure($()), ex -> new Failed<>(ex.toString()))
                    );

                    cachedTweets = cachedTweets.put(citizenName, remoteTweets);
                });
        return cachedTweets.get(citizenName).getOrElse(new NotRequestedYet<>());
    }

    private static Future<List<Tweet>> getTweetsAsync(String citizenName) {
        Future<? extends Citizen> futureCitizen = DbClient.getCitizenByName(citizenName);
        return futureCitizen.flatMap(DbClient::getTweetsFor);
    }
}
